docs/tutorials/RNotes/notes.md

```diff
@@ -0,0 +1,678 @@
+        provider.navigation?.push(controller: self.noteModification(note: nil))
+      },
+      modifyNote: { note in
+        provider.navigation?.push(controller: self.noteModification(note: note))
+      })
+    return MainController(reactions: reactions)
+  }
+}
+```
+
+Now that we have `Wireframe` pushing controllers when user decides to create a new or modify old note, we need to add such support to `MainController.swift`.
+
+Adding this snippet will create a button on the right side of the navigation bar and tapping on it will call our reaction.
+
+```swift
+// inside MainController's afterInit() method
+navigationItem.rightBarButtonItem = UIBarButtonItem(title: "New", style: .plain) { [reactions]
+  reactions.newNote()
+}
+```
+
+But the user might as well tap a note that is already created, for that we will use the `act(on:)` method.
+
+```swift
+// inside MainController
+override func act(on action: PlainTableViewAction<NoteCell>) {
+  switch action {
+  case .selected(let note):
+    reactions.modifyNote(note)
+  case .refresh, .rowAction(_, _):
+    break
+  }
+}
+```
+
+`act(on:)` allows the controller to do something every time `RootView` emits an `action` either through the `actions` observable or using the `perform(action:)` method.
+
+This concludes our transitioning to editing/creating notes and returning back (navigation controller takes care of returning for us). However we still do not have note-saving capability, everything returns to the same state as it was before the moment we leave the editing screen. That leads us to the next part.
+
+### Part 5: Saving Our Notes
+Any type of a request for data should be going through **Services**. We're going to only save our notes locally, although this abstraction allows you to switch saving and loading from `UserDefaults` to the cloud only changing the `Service`, nothing else.
+
+We'll create only one service, because our application is not very big, but there are usually many `Services`, just as there are plenty of `Wireframe`s.
+
+First of all, we need to make sure our `Note` can be serialized and deserialized. All we need to do is make `Note` conform to `Codable`, Swift takes care of everything else.
+
+```swift
+struct Note: Codable { // Note internal structure follows
+```
+
+Having done that, create a new file `NoteService.swift` in the `Services` folder. Add this code inside:
+
+```swift
+final class NoteService {
+  private let encoder = JSONEncoder()
+  private let decoder = JSONDecoder()
+
+  func loadNotes() throws -> [Note] {
+    guard let data = UserDefaults.standard.data(forKey: "notes") else { return [] }
+    return try decoder.decode(Array<Note>.self, from: data)
+  }
+
+  func save(note: Note) throws {
+    var notes = try loadNotes()
+    if let index = notes.index(where: { $0.id == note.id }) {
+      notes[index] = note
+    } else {
+      notes.append(note)
+    }
+    try save(notes: notes)
+  }
+
+  func save(note: Note) {
+    var notes = allNotes
```

Comment:
What is `allNotes`?

Reply:
Fixed, it was old implementation leaking. Good catch.